src/inmanta/server/services/notificationservice.py

```diff
@@ -70,7 +70,17 @@ async def compile_done(self, compile: data.Compile) -> None:
             failed_pull_stage = next(
                 (report for report in reports if report["name"] == "Pulling updates" and report["returncode"] != 0), None
             )
-            if failed_pull_stage:
+            if compile.notify_failed_compile == False:
+                return
+            elif compile.notify_failed_compile:
+                await self.notify(
+                    compile.environment,
+                    title="Compilation failed",
+                    message=compile.failed_compile_message if compile.failed_compile_message else "A compile has failed",
+                    severity=const.NotificationSeverity.error,
+                    uri=f"/api/v2/compilereport/{compile.id}",
```

Comment:
The notifications mentioned below this one in the `elif` sequence have a more specific message. Putting this condition at the start might cause a specific notification message to be overriden by a generic one, which is a step backwards in my opinion. An attempt to improve it:

```
if compile.notify_failed_compile and compile.failed_compile_message is not None:
    # Use specific message provided in request
    await self.notify()
elif failed_pull_stage:
    await self.notify()
elif compile.do_export:
    await self.notify()
elif compile.notify_failed_compile:
    # Send notification with generic message as fallback
    await self.notify()
``` 
I am not entirely sure what is best here. It might even be better to remove the first if condition. That way we force a specific notification message over the one provided in the request. @sanderr Do you have an opinion on this?

Reply:
I think you're suggestion is an improvement of behavior. The logic is slightly more complicated but I think that's an acceptable price to pay. About dropping the first condition, I had thought about this before but I concluded that it could be counterintuitive if the user requests a specific message and we override it in certain circumstances. I don't feel too strongly about this but I'd prefer to keep the first condition.